test/hotspot/jtreg/serviceability/jvmti/events/NotifyFramePopStressTest/libNotifyFramePopStressTest.cpp

```diff
@@ -0,0 +1,199 @@
+      deallocate(jvmti, jni, name);
+    }
+  }
+  deallocate(jvmti, jni, csig);
+  deallocate(jvmti, jni, name);
+}
+
+static
+jint Agent_Initialize(JavaVM *jvm, char *options, void *reserved) {
+  jint res;
+  jvmtiError err;
+
+  res = jvm->GetEnv((void **) &jvmti, JVMTI_VERSION_9);
+  if (res != JNI_OK || jvmti == nullptr) {
+    LOG("GetEnv(JVMTI_VERSION_9) failed error(%d)", res);
+    return JNI_ERR;
+  }
+  err = jvmti->GetPotentialCapabilities(&caps);
+  check_jvmti_error(err, "Agent: GetPotentialCapabilities failed");
+
+  err = jvmti->AddCapabilities(&caps);
+  check_jvmti_error(err, "Agent: AddCapabilities failed");
+
+  err = jvmti->GetCapabilities(&caps);
+  check_jvmti_error(err, "Agent: GetCapabilities failed");
+
+  if (caps.can_generate_frame_pop_events) {
+    callbacks.FramePop = &FramePop;
+    err = jvmti->SetEventCallbacks(&callbacks, sizeof(callbacks));
+    check_jvmti_error(err, "Agent: SetEventCallbacks failed");
+  }
+  return JNI_OK;
+}
+
+JNIEXPORT jboolean JNICALL
+Java_NotifyFramePopStressTest_canGenerateFramePopEvents(JNIEnv *env, jclass cls) {
+  return caps.can_generate_frame_pop_events ? JNI_TRUE : JNI_FALSE;
+}
+
+JNIEXPORT void JNICALL
+Java_NotifyFramePopStressTest_setFramePopNotificationMode(JNIEnv *env, jclass cl, jthread thread) {
+  set_event_notification_mode(jvmti, env, JVMTI_ENABLE, JVMTI_EVENT_FRAME_POP, thread);
+}
+
+/*
+ * Call NotifyFramePop on the current frame.
+ */
+JNIEXPORT jboolean JNICALL
+Java_NotifyFramePopStressTest_notifyFramePop(JNIEnv *jni, jclass cls, jthread thread) {
+  jmethodID method;
+  jlocation loc;
+  char* name;
+  jvmtiError err;
+  jboolean isMain;
+
+  err = jvmti->GetFrameLocation(thread, 0, &method, &loc);
+  check_jvmti_status(jni, err, "notifyFramePop: Failed in JVMTI GetFrameLocation");
+
+  name = get_method_name(jvmti, jni, method);
+
+  // We only want to do a NotifyFramePop once for the main method. The sole purpose is
+  // to force the thread into interpOnly mode, which seems to help the test's timing
+  // in a way that makes it more likely to reproduce the issue.
+  isMain = (strcmp(name, "main") == 0);
+  if (isMain) {
+    if (seenMain) {
+      deallocate(jvmti, jni, name);
+      return JNI_FALSE; // Only do NotifyFramePop once for main()
+    } else {
+      seenMain = JNI_TRUE;
+    }
+  }
+
+  err= jvmti->NotifyFramePop(thread, 0);
+  if (err == JVMTI_ERROR_OPAQUE_FRAME || err == JVMTI_ERROR_DUPLICATE) {
+    //LOG("\nNotifyFramePop for method %s returned acceptable error: %s\n", name, TranslateError(err));
+    deallocate(jvmti, jni, name);
+    return JNI_FALSE;
+  }
+  check_jvmti_status(jni, err, "notifyFramePop: Failed in JVMTI notifyFramePop");
+  LOG("\nNotifyFramePop called for method %s\n", name);
+
+  if (isMain) {
+    LOG("notifyFramePop not counting main method\n");
+    deallocate(jvmti, jni, name);
+    return JNI_FALSE;
+  } else {
+    char* old_notify_method = (char*)last_notify_method;
+    last_notify_method = name;
```

Comment:
I don't think this adds reliability.
Control thread suspends main test thread before call this method. AFAIU suspend cannot complete before FramePop callback is completed.

Reply:
Yes, this is all very synchronous. The control thread suspends the thread, we call NotifyFramePop on the thread, the control thread resumes the thread and then waits for confirmation of the FRAME_POP before repeating the process. I'm not sure what motivated the change to add old_notify_method. Maybe there was some other issue that was misunderstood.

Reply:
Control thread waits until pop_count is updated and FramePop callback updates it at the beginning, so control thread can start next iteration (i.e. call SuspentThread) before FramePop is completed.
But my understanding that SuspentThread cannot return before FramePop completed, so there is no synchronization issues here

Reply:
Actually I'm not so sure of your last statement. I think while in the FramePop callback the thread can do things that allow it to be suspended. So if the control thread has already detected that FramePop callback has been called, it may move on to the next iteration before the FramePop callback completes.

Reply:
Agreed, thanks.
The suspend should happen in the `ThreadInVMfromJava()` destructor that is a part of the `JRT_BLOCK` macro.

Reply:
> I'm not sure what motivated the change to add old_notify_method. Maybe there was some other issue that was misunderstood.

The mach5 test runs hit some strange crashes in `deallocate()` and made a wrong conclusion.

Reply:
> So if the control thread has already detected that FramePop callback has been called, it may move on to the next iteration before the FramePop callback completes.

This is important point to understand about the test.